src/main/scala/firrtl/transforms/TopWiring.scala

```diff
@@ -0,0 +1,247 @@
+            case None => sourceMap(currentmodule.name) = Seq((ComponentName(w.name,currentmodule), tpe, isport ,Seq[String](w.name)))
+          }
+          //sourceMap(currentmodule.name) +: (ComponentName(w.name,currentmodule), tpe, isport ,Seq[String](currentmodule.name))
+      }
+      state // Return argument unchanged (ok because DefWire has no Statement children)
+    // If not, apply to all children Statement
+    case _ => state
+  }
+
+
+
+
+
+
+
+  /** Create a map of Module name to target wires under this module
+    *
+    * These paths are relative but cross module (they refer down through instance hierarchy)
+    */
+  private def getSourcesMap(state: CircuitState): Map[String, Seq[(ComponentName, Type, Boolean, InstPath)]] = {
+    val sSourcesModNames = getSourceModNames(state)
+    val sSourcesNames = getSourceNames(state)
+     
+
+    val cMap = WiringUtils.getChildrenMap(state.circuit).toMap
+    //val digraph = DiGraph(cMap.mapValues(v => v.map(_._2).toSet))
+    //val topSort = digraph.linearize.reverse
+    //previous 2 line replaced by the following single line
+    val topSort = (new firrtl.analyses.InstanceGraph(state.circuit)).moduleOrder.reverse
+
+    // Map of component name to relative instance paths that result in a debug wire
+    val sourcemods: mutable.Map[String, Seq[(ComponentName, Type, Boolean, InstPath)]] =
+      mutable.Map(sSourcesModNames.map(_ -> Seq()): _*)
+
+    state.circuit.modules.foreach { m => m map getSourceTypes(sSourcesNames, sourcemods, ModuleName(m.name, CircuitName(state.circuit.main)) , state) }
+    state.circuit.modules.foreach { m => m.ports.foreach { p => Seq(p) map getSourceTypesPorts(sSourcesNames, sourcemods, ModuleName(m.name, CircuitName(state.circuit.main)) , state) }}
+
+    for (mod <- topSort) {
+      val seqChildren: Seq[(ComponentName,Type,Boolean,InstPath)] = cMap(mod.name).flatMap { case (inst, module) => 
+        sourcemods.get(module).map( _.map { case (a,b,c,path) => (a,b,c, inst +: path)})
+      }.flatten
+      if (seqChildren.nonEmpty) {
+        sourcemods(mod.name) = seqChildren
+      }
+    }
+
+    sourcemods.toMap
+  }
+
+
+
+  /** Process a given DefModule
+    *
+    * For Modules that contain or are in the parent hierarchy to modules containing target wires
+    * 1. Add ports for each target wire this module is parent to
+    * 2. Connect these ports to ports of instances that are parents to some number of target wires
+    */
+  // TODO Make code more clear
+  private def onModule(prefix: String, sources: Map[String, Seq[(ComponentName, Type, Boolean, InstPath)]])
+                      (module: DefModule): DefModule = {
+    sources.get(module.name) match {
+      case Some(p) => 
+        val newPorts = p.map{ case (ComponentName(cname,_), tpe, _ , path) => Port(NoInfo, prefix + path.mkString("_"), Output, tpe)}
```

Comment:
I like the use of `InstPath` to determine the name (you get an understanding of where the signal is going), however, this can still collide with other modules components, yes? You can use a `Namespace` here to make sure that the name is valid and then either get a new one or fail if you really wanted that name.

Reply:
Aren't instance names unique? if an instance name is unique, than the InstPath must be unique as well, isn't it? Can you give me an example of a collision?

Reply:
Instance names are guaranteed to be unique. However, there's nothing preventing a user from defining an existing wire/port/reg in the design with one of these names. In your first test case, you create the top-level port `output topwiring_a1_b1_c1_x: UInt<1>`. However, the user could have created this signal already defined `input topwiring_a1_b1_c1_x: Clock`. This _shouldn't_ happen, but the user should be able to name (or Chisel/earlier FIRRTL passes) should be able to name signals whatever they want without having to worry about this transform generating conflicting names.